test/test_autograd.py

```diff
@@ -491,6 +500,91 @@ def fn(x):
         with self.assertRaisesRegex(RuntimeError, "expects an grad_fn"):
             torch._C._will_engine_execute_node(out)
 
+    def test_custom_function_setup_context_simple(self):
+        class MySquare(Function):
+            @staticmethod
+            def forward(x):
+                return x ** 2
+
+            @staticmethod
+            def setup_context(ctx, outputs, x):
+                ctx.save_for_backward(x)
+
+            @staticmethod
+            def backward(ctx, gO):
+                x, = ctx.saved_tensors
+                return gO * 2 * x
+
+        with enable_autograd_function_extension():
+            x = torch.randn([], requires_grad=True)
+            y = MySquare.apply(x)
+            gx, = torch.autograd.grad(y, x)
+            self.assertEqual(gx, 2 * x)
+
+    def test_custom_function_setup_context_multi_output(self):
+        # Multiple outputs with some non-Tensor outputs.
+        class MySquare(Function):
+            @staticmethod
+            def forward(x):
+                two_x = x.item() * 2
+                return x ** 2, two_x
+
+            @staticmethod
+            def setup_context(ctx, outputs, x):
+                _, two_x = outputs
+                ctx.two_x = two_x
+
+            @staticmethod
+            @once_differentiable
+            def backward(ctx, gO, _):
+                return gO * ctx.two_x
+
+        with enable_autograd_function_extension():
+            x = torch.randn([], requires_grad=True)
+            y, _ = MySquare.apply(x)
+            gx, = torch.autograd.grad(y, x)
+            self.assertEqual(gx, 2 * x)
+
+    def test_custom_function_setup_context_multi_input(self):
+        class MyReshape(Function):
+            @staticmethod
+            def forward(x, shape, scale_forward, scale_backward):
+                return x.reshape(shape) * scale_forward
+
+            @staticmethod
+            def setup_context(ctx, outputs, x, shape, copy_forward, scale_backward):
```

Comment:
Another option is to have both outputs and inputs non-expanded. Not sure how much better it is because users would always be able to just do `setup_context(ctx, outputs, *inputs)` anyway. It does seem more consistent and would allow inputs to be before outputs.

Reply:
> unexpanded inputs can be consistent across the various staticmethods

What happens right now if they aren't consistent? As an example, I could see a user trying to rewrite their existing autograd.Function and accidentally ending up with

```python
class Foo(torch.autograd.Function)
  @staticmethod
  def forward(ctx, x, y):
    ctx.save_for_backward(x, y) # a vestiage from the old way
    return torch.tensor(to_numpy(x) * to_numpy(y), device=x.device)

  @staticmethod
  def setup_context(ctx, outputs, x, y):
    ctx.save_for_backward(x, y)
```
and want to check what the error message looks like